### PR TITLE
Intercept app data for DTLS earlier

### PR DESCRIPTION
**Pull request by willardf**
We were intercepting bytes at WriteBytesToConnection, which is too late to avoid starting resend timers for reliable messages. Also I'm wondering if the messages might accidentally be encrypted wrong? Like using the previous cypher if we have to renegotiate in the middle of a connection.

This is a draft because there is one failing test that needs to be fixed, but I wanted to get some feedback on what's happening and the logic of said failing test.

TestMalformedConnectionData is the only failing test because the DTLS was relying on the UDP hello packet timing out to fail the Connect. While it's important that after Connect times out, we need to stop sending packets and set the connection to NotConnecting (failure to connect). It seems... double bad? that we were relying on that kind of failure when the DTLS handshake itself has nothing to do with the underlying hello packets.

In short, I'm going to continue to poke around for a good way to have DTLS bubble up a failure to connect, but I wanted to flag this as kind of a weird thing. It's very elegant that DTLS sits on top of the UDP layer and doesn't care about the underlying send or retry mechanisms, but it also means that it needs to recreate some of those mechanisms, which it might not be doing right now.

Edit + Ready for Review: Okay, so adding timeout checks on the handshake resends seems pretty adequate. The one test is fixed, but I'm mostly left with the feeling that the testing is inadequate, but things are okay other than my anxiety.

### CODE REVIEW

**Comment by willardf** [comment]
As I'm trying to figure out where to insert a timeout, I'm realizing that it's unlikely a connection can properly disconnect due to epoch negotiation ever? So figuring this out *might* fix a bug with established connections as well. Still trying to sort that bit out, though.

**Comment by willardf** [comment]
Okay, I have a fix that seems valid, but I'm still not sure if the epoch can actually ever be renegotiated in the middle of a connection. It's a good thing if no because it means fewer cornercases and I suspect there's no really good reason to. It's just a hole in my understanding.

File: Hazel/Dtls/DtlsUnityConnection.cs [hunk 1]
```diff
@@ -213,23 +225,34 @@ protected override void ResendPacketsIfNeeded()
                     DateTime now = DateTime.UtcNow;
                     if (now >= this.nextEpoch.NextPacketResendTime)
                     {
-                        switch (this.nextEpoch.State)
+                        double negotationDuration = (now - this.nextEpoch.NegotationStartTime).TotalMilliseconds;
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsUnityConnection.cs:228]: `negotiationDuration`

File: Hazel/Dtls/DtlsUnityConnection.cs [hunk 2]
```diff
@@ -62,7 +64,9 @@ struct NextEpoch
 
             public ulong NextOutgoingSequence;
 
+            public DateTime NegotationStartTime;
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsUnityConnection.cs:67]: `NegotiationStartTime`

File: Hazel/Dtls/DtlsUnityConnection.cs [hunk 3]
```diff
@@ -213,23 +225,34 @@ protected override void ResendPacketsIfNeeded()
...
+                        if ((this.ResendLimit > 0 && this.nextEpoch.PacketResendCount > this.ResendLimit)
+                            || negotationDuration > this.DisconnectTimeout)
                         {
-                            case HandshakeState.ExpectingServerHello:
-                            case HandshakeState.ExpectingCertificate:
-                            case HandshakeState.ExpectingServerKeyExchange:
-                            case HandshakeState.ExpectingServerHelloDone:
-                                this.SendClientHello();
-                                break;
+                            this.DisconnectInternal(HazelInternalErrors.DtlsNegotiationFailed, $"DTLS negotiation failed after {this.nextEpoch.PacketResendCount} resends ({(int)negotationDuration} ms).");
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsUnityConnection.cs:234]: Great! I think this solves one of the error cases that've been working on @nommiin 

We might want to expand on this a bit to detect a "server never reached" case where no packets were ever received, and I think we can work from this

File: Hazel/Dtls/DtlsUnityConnection.cs [hunk 4]
```diff
@@ -213,23 +225,34 @@ protected override void ResendPacketsIfNeeded()
                     DateTime now = DateTime.UtcNow;
                     if (now >= this.nextEpoch.NextPacketResendTime)
                     {
-                        switch (this.nextEpoch.State)
+                        double negotationDuration = (now - this.nextEpoch.NegotationStartTime).TotalMilliseconds;
+                        this.nextEpoch.PacketResendCount++;
+
+                        if ((this.ResendLimit > 0 && this.nextEpoch.PacketResendCount > this.ResendLimit)
+                            || negotationDuration > this.DisconnectTimeout)
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsUnityConnection.cs:232]: Assuming DisconnectTimeout is already existing and used by consumers of the library, so I see reason to not change, but, just in case I'd recommend that the units are baked into the name, ie `DisconnectTimeoutMilliseconds` - I find this helps a lot from preventing unit mismatch mistakes

(And perhaps we can make it clearer here for updating to `negotiationDurationMilliseconds`)

File: Hazel/Dtls/DtlsUnityConnection.cs [hunk 5]
```diff
@@ -89,7 +100,7 @@ struct NextEpoch
         private NextEpoch nextEpoch;
         private TimeSpan handshakeResendTimeout = TimeSpan.FromMilliseconds(200);
 
-        private readonly List<ByteSpan> queuedApplicationData = new List<ByteSpan>();
+        private readonly ConcurrentQueue<QueuedAppData> queuedApplicationData = new ConcurrentQueue<QueuedAppData>();
```
Comment (Mukikaizoku) [inline Hazel/Dtls/DtlsUnityConnection.cs:103]: Had a question here---is this part of the library run off of the Unity main thread? (or is that a possible use case?). Just checking if the use of Concurrent was intentional or if we can switch to a standard Queue

**Review by Mukikaizoku — CHANGES_REQUESTED** [review]
Solution looks great to me and I think this will solve the issue we noticed with the concurrent SendHello, now that it is queued/delayed in this DTLS layer!

Had a few minor questions and requests!

**Comment by Mukikaizoku** [comment]
> TestMalformedConnectionData is the only failing test because the DTLS was relying on the UDP hello packet timing out to fail the Connect. While it's important that after Connect times out, we need to stop sending packets and set the connection to NotConnecting (failure to connect). It seems... double bad? that we were relying on that kind of failure when the DTLS handshake itself has nothing to do with the underlying hello packets.

Yeah, going through the code today, had noticed that the DTLS layer was relying on the SendHello timeout to timeout an infinite DTLS connection attempt. Agreed that the logic of the layers shouldn't be intertwined---so nice that this PR fixes that here.

 
> In short, I'm going to continue to poke around for a good way to have DTLS bubble up a failure to connect, but I wanted to flag this as kind of a weird thing. It's very elegant that DTLS sits on top of the UDP layer and doesn't care about the underlying send or retry mechanisms, but it also means that it needs to recreate some of those mechanisms, which it might not be doing right now.

The decorator setup for the DTLS layer is indeed elegant here and nicely isolates the concerns. While not 100% sure it was implied, I do agree that having to recreate the same types of logic (timeouts, reliability/resends) in this layer seems not ideal?

I think the changes at hand are great, but for food for thought, an alternative approach may be to insert DTLS in the "middle":

[Application calls Connect/Send] -> [Connection layer] -> [DTLS layer] -> [Reliability layer] -> [Send layer]

Here, DTLS can send out packets and they are handled like any reliable packets. The Reliability layer would detect timeouts and bubble that back to whatever called it (DTLS or Connection , and DTLS will clean its state and bubble back to Connection  layer)---and ultimately Connection layer handles the disconnect report.

The equivalent arrows drawn for the current setup, with the inheritance hierarchy, would likely weave in and out of each of the layers, and I think that's why it is difficult to reuse the reliability and timeout logic.

File: Hazel/Dtls/DtlsUnityConnection.cs (hunk 5 above)
Comment (willardf) [reply Hazel/Dtls/DtlsUnityConnection.cs:103]: Yeah, this is run off the Unity main thread, but because of the syncLock, technically we could use a normal Queue. Haha, I just like the syntax of the TryDequeue better. Maybe I should just make an extension method for the normal Queue...

**Comment by willardf** [comment]
We talked offline about how the organization of Hazel is very heavily inherited and how that makes it difficult to follow the data and test code units. It is possible in the future Hazel will undergo a large refactor to componentize sections like DTLS or Reliability for these reasons, but it'll be a large undertaking that will be hard to prioritize.

File: Hazel/Dtls/DtlsUnityConnection.cs (hunk 5 above)
Comment (Mukikaizoku) [reply Hazel/Dtls/DtlsUnityConnection.cs:103]: Haha yeah, I feel you on the TryDequeue---every time I use a normal queue I feel it missing!

And cool, I don't think it'll be an issue either way, but might be good measure to change / add the extension method

**Review by Mukikaizoku — APPROVED** [review]
Thank you for the changes---they look good! Possibly update the queue per discussion in that thread, but that doesn't seem to be an issue to me so will approve!

**Comment by willardf** [comment]
So one of the tests was consistently failing, but I couldn't figure out if it was UDP or DTLS, and when I split things to figure it out, the test started passing. Never repro'd locally, so I guess I'm good to go?

File: Hazel/Extensions.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections.Generic;
+
+namespace Hazel
+{
+    public static class Extensions
+    {
+        public static bool TryDequeue<T>(this Queue<T> self, out T item)
```
Comment (Mukikaizoku) [inline Hazel/Extensions.cs:7]: Sweeeet!

**Comment by Mukikaizoku** [comment]
Hahaha yeah that is strange---it wasn't obvious from the code either. And yeah, I think you're good to go!

**Review by Mukikaizoku — APPROVED** [review]
Approving the updates!
